File: Leetcode/0211 Design Add and Search Words Data Structure.cpp
// implemented trie

class WordDictionary {
    struct Node {
        unordered_map<char, Node*> children;
        bool is_end_of_the_word;
        
        Node() {
            is_end_of_the_word = false;
        }
        
        bool contains_child(char child_char) {
            return children.find(child_char) != children.end();
        }
    };
    Node root;

    bool search(Node *node, const string &word, int start_index) {
        const char &c = word[start_index]; 
        if(start_index == word.size()) {
            return node->is_end_of_the_word;
        }
        bool exists = false;
        if(c == '.') {
            for(const auto &[key, value] : node->children) {
                exists |= search(value, word, start_index + 1);
            }
        } else if(node->contains_child(c)){
            exists |= search(node->children[c], word, start_index + 1);
        }
        return exists;
    }
public:
    WordDictionary() {
        
    }
    
    void addWord(string word) {
        Node *curr_node = &root;
        int n = word.size();
        for(int i = 0;i < n; ++i) {
            const char &c = word[i];
            if(!curr_node->contains_child(c)) {
                curr_node->children[c] = new Node;
            }
            curr_node = curr_node->children[c];                
            if(i == (n - 1)) {
                curr_node->is_end_of_the_word = true;
            }
        }
    }
    
    bool search(string word) {
        return search(&root, word, 0);
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */

File: Leetcode/0211 Design Add and Search Words Data Structure.java
class WordDictionary {
    private final Node root;

    public WordDictionary() {
        root = new Node();
        root.isWordEnd = true;
    }

    public void addWord(String word) {
        var currNode = root;
        for (int i = 0; i < word.length(); ++i) {
            int charIndex = word.charAt(i) - 'a';
            if (currNode.children[charIndex] == null) {
                currNode.children[charIndex] = new Node();
            }
            currNode = currNode.children[charIndex];
        }
        currNode.isWordEnd = true;
    }

    public boolean search(String word) {
        return search(0, root, word);
    }

    private boolean search(int index, Node node, String word) {
        if (node == null) {
            return false;
        }
        if (index == word.length()) {
            return node.isWordEnd;
        }
        char ch = word.charAt(index);
        if (ch != '.') {
            return search(index + 1, node.children[ch - 'a'], word);
        }
        for (int charIndex = 0; charIndex < ALPHABET_SIZE; ++charIndex) {
            if (search(index + 1, node.children[charIndex], word)) {
                return true;
            }
        }
        return false;
    }

    private static final int ALPHABET_SIZE = 26;

    private static class Node {
        boolean isWordEnd;
        Node[] children = new Node[ALPHABET_SIZE];
    }
}